oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/tree/impl/OrderedChildnameIterator.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.tree.impl;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ *  Return the childrenNames in the order defined by the orderedChildren iterator, and merges it
+ *  with the existing children defined by allChildren.
+ *
+ *  This implementation focuses on being as lazy as possible; especially consuming the
+ *  allChildren iterator can be slow.
+ */
+
+public class OrderedChildnameIterator implements Iterator<String>{
+
+    final Iterator<String> orderedChildren;
+    final Iterator<String> allChildren;
+
+    private String nextResult;
+
+    private final List<String> nonOrderedChildren = new ArrayList<>();
+    private Iterator<String> nonOrderedChildrenIterator = null;
+
+    public OrderedChildnameIterator (Iterable<String> orderedChildren, Iterable<String> allChildren) {
+        this.orderedChildren = orderedChildren == null ? Collections.emptyIterator() : orderedChildren.iterator();
+        this.allChildren = allChildren.iterator();
+        nextResult = getNextElement();
+    }
+
+    private String getNextElement() {
+        if (orderedChildren.hasNext()) {
+            String elem = getNextOrderedChild();
+            if (elem != null) {
+                return elem;
+            }
+        }
+        // if the flow comes here, all orderedChildren have already been consumed, and the
+        // nonOrderedChildren list is no longer changed, so it's safe to create the iterator here
+        if (nonOrderedChildrenIterator == null) {
+            nonOrderedChildrenIterator = nonOrderedChildren.iterator();
+        }
+        // return all children which have already been read into the nonOrderedChildren list
+        if (nonOrderedChildrenIterator.hasNext()) {
+            return nonOrderedChildrenIterator.next();
+        }
+        // return all children which have not been consumed from the allChildren iterator
+        if (allChildren.hasNext()) {
+            return allChildren.next();
+        }
+        // all iterators consumed, no children anymore
+        return null;
+    }
+
+    /**
+     * Consume the next element from the orderedChild list and validates that it's actually present
+     * @return the next ordered child or  {code null} if all ordered children have already been returned
+     */
+    @SuppressWarnings("java:S2589")
+    private String getNextOrderedChild() {
+        String current = null;
+        // check that this element is actually present in the allChildren iterable
+        while (current == null && orderedChildren.hasNext()) {
```

Comment:
my IDE complains the current is always null and thus the check is redundant.

Reply:
that's a good spot, I even put in an explicit ignore, because I mis-understood it.